### PR TITLE
Typing/added typing to httptools_impl.py module

### PR DESCRIPTION
**Pull request by jkklapp**
Part of #998

### CODE REVIEW

**Comment by jkklapp** [comment]
wip

**Comment by jkklapp** [comment]
@Kludex I added some `type: ignore` lines after spending a bit with each. It would be good to have some help from you or others to try and fix them though.

**Comment by jkklapp** [comment]
@Kludex bump :)

File: uvicorn/protocols/http/httptools_impl.py [hunk 1]
```diff
@@ -38,38 +43,38 @@ def _get_status_line(status_code):
...
 
-    def resume_writing(self):
+    def resume_writing(self) -> None:
         if self.write_paused:
             self.write_paused = False
             self._is_writable_event.set()
 
 
-async def service_unavailable(scope, receive, send):
+async def service_unavailable(scope: Scope, receive: Callable, send: Callable) -> None:
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]: The idea is to follow `asgiref`. One of my PRs already have the types for `receive` and `send`.

File: uvicorn/protocols/http/httptools_impl.py [hunk 2]
```diff
@@ -84,7 +89,12 @@ async def service_unavailable(scope, receive, send):
 
 
 class HttpToolsProtocol(asyncio.Protocol):
-    def __init__(self, config, server_state, _loop=None):
+    def __init__(
+        self,
+        config: Config,
+        server_state: ServerState,
+        _loop: asyncio.AbstractEventLoop = None,
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]:
```suggestion
        _loop: Optional[asyncio.AbstractEventLoop] = None,
```

File: uvicorn/protocols/http/httptools_impl.py [hunk 3]
```diff
@@ -110,22 +120,22 @@ def __init__(self, config, server_state, _loop=None):
...
-        self.server = None
-        self.client = None
-        self.scheme = None
-        self.pipeline = []
+        self.transport: asyncio.Transport
+        self.flow: FlowControl
+        self.server: Optional[Tuple[str, int]] = None
+        self.client: Optional[Tuple[str, int]] = None
+        self.scheme: str
+        self.pipeline: list = []
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py:140]: Too generic.

File: uvicorn/protocols/http/httptools_impl.py [hunk 4]
```diff
@@ -110,22 +120,22 @@ def __init__(self, config, server_state, _loop=None):
...
         # Per-request state
-        self.url = None
-        self.scope = None
-        self.headers = None
+        self.url: Optional[str] = None
+        self.scope: dict = {}
+        self.headers: list = []
         self.expect_100_continue = False
-        self.cycle = None
+        self.cycle: RequestResponseCycle = None  # type: ignore
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py:147]: Why ignoring?
```suggestion
        self.cycle: Optional[RequestResponseCycle] = None  # type: ignore
```

File: uvicorn/protocols/http/httptools_impl.py [hunk 5]
```diff
@@ -110,22 +120,22 @@ def __init__(self, config, server_state, _loop=None):
...
+        self.scheme: str
+        self.pipeline: list = []
 
         # Per-request state
-        self.url = None
-        self.scope = None
-        self.headers = None
+        self.url: Optional[str] = None
+        self.scope: dict = {}
+        self.headers: list = []
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py:145]: Is it `List[Tuple[str]]`?

File: uvicorn/protocols/http/httptools_impl.py [hunk 6]
```diff
@@ -110,22 +120,22 @@ def __init__(self, config, server_state, _loop=None):
...
+        self.client: Optional[Tuple[str, int]] = None
+        self.scheme: str
+        self.pipeline: list = []
 
         # Per-request state
-        self.url = None
-        self.scope = None
-        self.headers = None
+        self.url: Optional[str] = None
+        self.scope: dict = {}
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]: Too generic.

File: uvicorn/protocols/http/httptools_impl.py [hunk 7]
```diff
@@ -110,22 +120,22 @@ def __init__(self, config, server_state, _loop=None):
...
+        self.headers: list = []
         self.expect_100_continue = False
-        self.cycle = None
+        self.cycle: RequestResponseCycle = None  # type: ignore
 
     # Protocol interface
-    def connection_made(self, transport):
+    def connection_made(self, transport: asyncio.Transport) -> None:  # type: ignore
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py:150]: Why ignoring?

File: uvicorn/protocols/http/httptools_impl.py [hunk 8]
```diff
@@ -262,7 +272,7 @@ def on_headers_complete(self):
 
         existing_cycle = self.cycle
         self.cycle = RequestResponseCycle(
-            scope=self.scope,
+            scope=self.scope,  # type: ignore
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]: What's the issue?

File: uvicorn/protocols/http/httptools_impl.py [hunk 9]
```diff
@@ -389,11 +399,11 @@ def __init__(
         # Response state
         self.response_started = False
         self.response_complete = False
-        self.chunked_encoding = None
+        self.chunked_encoding: Optional[bool] = None
         self.expected_content_length = 0
 
     # ASGI exception wrapper
-    async def run_asgi(self, app):
+    async def run_asgi(self, app: Callable) -> None:
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]: Too generic. One of my PRs have this already, or even the current one already has? Not sure.

File: uvicorn/protocols/http/httptools_impl.py [hunk 10]
```diff
@@ -435,7 +445,7 @@ async def send_500_response(self):
         )
 
     # ASGI interface
-    async def send(self, message):
+    async def send(self, message: dict) -> None:
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]: asgiref. One of my PRs already has the `Message` type.

File: uvicorn/protocols/http/httptools_impl.py [hunk 11]
```diff
@@ -544,7 +554,7 @@ async def send(self, message):
             msg = "Unexpected ASGI message '%s' sent, after response already completed."
             raise RuntimeError(msg % message_type)
 
-    async def receive(self):
+    async def receive(self) -> Dict:
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]: This is Message as well.

**Review by Kludex — COMMENTED** [review]
There are too many ignores, can you check those? I didn't comment on all of them. But I believe some of those are necessary because this file needs some others to be completed properly, am I right?

**Comment by jkklapp** [comment]
@Kludex I have been checking your PRs to add the correct types however Im getting mypy issues I am struggling to solve. Hence I added more lines with `type: ignore` :(

Is it useful if I paste some mypy outputs here?

**Comment by Kludex** [comment]
I don't know... Probably not to be honest. But we can try.

I've tried to crack this file before, but I felt like it depends on other stuff to be merged before :disappointed:

**Comment by Kludex** [comment]
@jkklapp We're now more active on those PRs.

We've choose to make use of [asgiref typing module](https://github.com/django/asgiref/blob/main/asgiref/typing.py) instead of the `_types.py` (which was removed). Do you mind updating this PR?

**Comment by Kludex** [comment]
@jkklapp Thanks for the efforts! I'll take the lead now. :smile:

**Pull request closed without merging**